### PR TITLE
WT-4715 Workloads will stall if old transaction or timestamp pinned by thread co-opted for eviction.

### PR DESCRIPTION
**Pull request by keithbostic**
@sueloverso, can you specifically review 0743600, those are changes to move your work in WT-5043 aside so I could grab the "set a timestamp at the API level" namespace. I also removed the `DEBUGGING` tag for the `WT_REF.hist.time_sec` field as we discussed, and renamed everything so it's hopefully obvious all of this work should be removed in 6 months.

@agorrod, can you specifically review the `__wt_txn_is_blocking_pin()` function in 3b29dd9? I'm not confident I'm making the correct checks before looking for the oldest pinned transaction ID, I don't know how that stuff fits together.

### CODE REVIEW

**Comment by keithbostic** [comment]
autobuild: retest this please

File: src/include/txn.i [hunk 1]
```diff
@@ -871,8 +871,12 @@ __wt_txn_begin(WT_SESSION_IMPL *session, const char *cfg[])
         if (session->ncursors > 0)
             WT_RET(__wt_session_copy_values(session));
 
-        /* Stall here if the cache is completely full. */
-        WT_RET(__wt_cache_eviction_check(session, false, true, NULL));
+        /*
+         * Stall here if the cache is completely full. It's possible for cache eviction to decide
+         * we're the problem (we might be pinning data in the cache), and return WT_ROLLBACK. The
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/include/txn.i]: It's been awhile, but I definitely saw this fail at some point. My recollection is we've already set our transaction ID by the time we get here, and so we can be the oldest transaction ID in the system?

If that doesn't sound right to you, let me know and I'll go back and figure it out.

**Comment by keithbostic** [comment]
@agorrod,
> I'm also a little nervous about the change. Do you think it'd make sense to switch some stress testing over here before merging?

I'm never going to object to more testing. I ran a bunch of `test/format` stress tests on it, but no sanitizer tests.

File: src/include/txn.i (hunk 1 above)
[the remark this answers is not in the document]
Comment (agorrod) [reply src/include/txn.i]: That sounds right, but I would like a different comment, maybe:
Stall here if the cache is completely full. We have allocated a transaction ID which makes it possible for eviction to decide we're contributing to the problem and return WT_ROLLBACK. The WT_SESSION.begin_transaction API can't return rollback, continue on.

**Comment by keithbostic** [comment]
@agorrod, 
> That sounds right, but I would like a different comment, maybe:

I've updated the comment in 30c4ac5.

We've got 4 good sanitizer runs, so I'll merge this as soon as it passes PR testing and switch the sanitizer runs back to develop.

**Pull request merged**